cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java

```diff
@@ -442,29 +442,29 @@ public void testQueryMetricsWithTags() throws Exception {
         + end);
 
     List<TimeSeriesResult> groupByResult =
-      ImmutableList.of(new TimeSeriesResult(ImmutableMap.of("flowlet", "counter"), 1),
-                       new TimeSeriesResult(ImmutableMap.of("flowlet", "splitter"), 3));
+      ImmutableList.of(new TimeSeriesResult(ImmutableMap.of("handler", "counter"), 1),
+                       new TimeSeriesResult(ImmutableMap.of("handler", "splitter"), 3));
 
     verifyGroupByResult(
       "/v3/metrics/query?" + getTags("yourspace", "WCount1", "WCounter") +
-        "&metric=system.reads&groupBy=flowlet&start=" + start + "&end="
+        "&metric=system.reads&groupBy=handler&start=" + start + "&end="
         + end, groupByResult);
 
-    groupByResult =
-      ImmutableList.of(new TimeSeriesResult(ImmutableMap.of("namespace", "myspace", "flowlet", "splitter"), 2),
-                       new TimeSeriesResult(ImmutableMap.of("namespace", "yourspace", "flowlet", "counter"), 1),
-                       new TimeSeriesResult(ImmutableMap.of("namespace", "yourspace", "flowlet", "splitter"), 4));
+    groupByResult = ImmutableList.of(
+        new TimeSeriesResult(ImmutableMap.of("namespace", "myspace", "handler", "splitter", "app", "WordCount1"), 2),
+        new TimeSeriesResult(ImmutableMap.of("namespace", "yourspace", "handler", "counter", "app", "WCount1"), 1),
+        new TimeSeriesResult(ImmutableMap.of("namespace", "yourspace", "handler", "splitter", "app", "WCount1"), 4));
 
     verifyGroupByResult(
       "/v3/metrics/query?metric=system.reads" +
-        "&groupBy=namespace&groupBy=flowlet&start=" + start + "&end=" + end, groupByResult);
+        "&groupBy=namespace&groupBy=handler&groupBy=app&start=" + start + "&end=" + end, groupByResult);
```

Comment:
why also groupBy=app?

Reply:
This is to workaround a problem in the metrics aggregate matching logic as in https://github.com/cdapio/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java#L360

In short, after switching to use Service instead of Flow, if the "groupBy=app" is not specified, the `findAggregate` logic will pick the aggregate rule for the system service, not user service, hence failing the test.

Reply:
got it.